### PR TITLE
Reworked logging to route log output through Python stderr

### PR DESCRIPTION
**Pull request by tpoterba**
- Removed SLF4J, everything goes directly through log4j now.
 - Removed the explicitly System.err.write calls inside info / warn.
 - Separated console logging and log file logging
 - Stripped the huge stack traces out of python errors; they are logged
   to the screen / cell through log4j.

### CODE REVIEW

**Comment by tpoterba** [comment]
Works with jupyter notebooks:
image

**Comment by tpoterba** [comment]
I just tested on Dataproc and it works there just fine.

File: python/hail/java.py [hunk 1]
```diff
@@ -119,3 +128,42 @@ def handle_py4j(func, *args, **kwargs):
...
+        SocketServer.TCPServer.__init__(self, server_address, RequestHandlerClass)
+
+
+def connect_logger(host, port):
+    server = None
+    tries = 0
+    max_tries = 25
+    while not server:
+        try:
+            server = SimpleServer((host, port), LoggingTCPHandler)
```
Comment (cseed) [inline python/hail/java.py]: You should put make this retry loop infinite, add a 1s sleep on failure, and put both the server creation and the call to serve_forever in the loop.

File: src/main/scala/is/hail/HailContext.scala [hunk 2]
```diff
@@ -174,25 +172,20 @@ object HailContext {
...
+      """
+        |Welcome to
         |     __  __     <>__
         |    / /_/ /__  __/ /
         |   / __  / _ `/ / /
         |  /_/ /_/\_,_/_/_/   version """.stripMargin + is.hail.HAIL_PRETTY_VERSION
-    println(welcomeMessage)
-    log.info(welcomeMessage)
+
+    info(welcomeMessage)
```
Comment (cseed) [inline src/main/scala/is/hail/HailContext.scala]: Do we need this here?  Just log the version.  I don't think we need to pretend the user is going to run through Scala anymore.

File: src/main/scala/is/hail/utils/Py4jUtils.scala [hunk 3]
```diff
@@ -67,6 +64,11 @@ trait Py4jUtils {
   def copyFile(from: String, to: String, hc: HailContext) {
     hc.hadoopConf.copy(from, to)
   }
+
+  def addSocketAppender(hostname: String, port: Int) {
+    val app = new StringSocketAppender(hostname, port, "%d{yyyy-MM-dd HH:mm:ss} %c{1}: %p: %m%n")
```
Comment (cseed) [inline src/main/scala/is/hail/utils/Py4jUtils.scala]: Use HailContext.logFormat.

File: src/main/scala/is/hail/utils/StringSocketAppender.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        case e: IOException =>
+          if (e.isInstanceOf[InterruptedIOException]) Thread.currentThread.interrupt()
+          LogLog.error("Could not close os.", e)
+      }
+      os = null
+    }
+    if (connector != null) { //LogLog.debug("Interrupting the connector.");
+      connector.interrupted = true
+      connector = null // allow gc
+
```
Comment (cseed) [inline src/main/scala/is/hail/utils/StringSocketAppender.scala]: Spurious newline.

File: src/main/scala/is/hail/utils/StringSocketAppender.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      try
+        os.close()
+      catch {
+        case e: IOException =>
+          if (e.isInstanceOf[InterruptedIOException]) Thread.currentThread.interrupt()
+          LogLog.error("Could not close os.", e)
+      }
+      os = null
+    }
+    if (connector != null) { //LogLog.debug("Interrupting the connector.");
```
Comment (cseed) [inline src/main/scala/is/hail/utils/StringSocketAppender.scala]: Remove comment or leave debug in.

File: src/main/scala/is/hail/utils/StringSocketAppender.scala [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    } catch {
+      case e: IOException =>
+        if (e.isInstanceOf[InterruptedIOException]) Thread.currentThread.interrupt()
+        var msg = "Could not connect to remote log4j server at [" + address.getHostName + "]."
+        if (reconnectionDelay > 0) {
+          msg += " We will try again later."
+          fireConnector() // fire the connector thread
+
+        }
+        else {
```
Comment (cseed) [inline src/main/scala/is/hail/utils/StringSocketAppender.scala]: Formatting.

**Comment by tpoterba** [comment]
addressed comments, back to you

**Pull request merged**